File: jre/src/test/java/com/azure/data/cosmos/serialization/hybridrow/unit/RecordIOUnitTests.java
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

package com.azure.data.cosmos.serialization.hybridrow.unit;

import com.azure.data.cosmos.core.Out;
import com.azure.data.cosmos.core.Reference;
import com.azure.data.cosmos.serialization.hybridrow.HybridRowVersion;
import com.azure.data.cosmos.serialization.hybridrow.MemorySpanResizer;
import com.azure.data.cosmos.serialization.hybridrow.Result;
import com.azure.data.cosmos.serialization.hybridrow.RowBuffer;
import com.azure.data.cosmos.serialization.hybridrow.io.RowReader;
import com.azure.data.cosmos.serialization.hybridrow.recordio.RecordIOStream;
import com.azure.data.cosmos.serialization.hybridrow.recordio.Segment;
import com.azure.data.cosmos.serialization.hybridrow.unit.customerschema.Address;

import java.nio.file.Files;
import java.util.ArrayList;

// TODO: C# TO JAVA CONVERTER: There is no preprocessor in Java:
///#pragma warning disable SA1401 // Fields should be private


// TODO: C# TO JAVA CONVERTER: Java annotations will not correspond to .NET attributes:
//ORIGINAL LINE: [TestClass][DeploymentItem(RecordIOUnitTests.SchemaFile, "TestData")] public class RecordIOUnitTests
public class RecordIOUnitTests {
    private static final int InitialRowSize = 0;
    private static final String SchemaFile = "TestData\\CustomerSchema.json";
    private Layout addressLayout;
    private Namespace ns;
    private LayoutResolver resolver;

    // TODO: C# TO JAVA CONVERTER: Java annotations will not correspond to .NET attributes:
    //ORIGINAL LINE: [TestMethod][Owner("jthunter")] public void LoadSchema()
    public final void LoadSchema() {
        LayoutResolver systemResolver = SystemSchema.LayoutResolver;
        Layout segmentLayout = systemResolver.Resolve(SystemSchema.SegmentSchemaId);
        assert segmentLayout.getName().equals("Segment");
        assert segmentLayout.getSchemaId().clone() == SystemSchema.SegmentSchemaId;

        Layout recordLayout = systemResolver.Resolve(SystemSchema.RecordSchemaId);
        assert recordLayout.getName().equals("Record");
        assert recordLayout.getSchemaId().clone() == SystemSchema.RecordSchemaId;
    }

    // TODO: C# TO JAVA CONVERTER: Java annotations will not correspond to .NET attributes:
    //ORIGINAL LINE: [TestInitialize] public void ParseNamespaceExample()
    public final void ParseNamespaceExample() {
        String json = Files.readString(RecordIOUnitTests.SchemaFile);
        this.ns = Namespace.Parse(json);
        this.resolver = new LayoutResolverNamespace(this.ns);
        this.addressLayout = this.resolver.Resolve(tangible.ListHelper.find(this.ns.getSchemas(), x -> x.Name.equals(
            "Address")).SchemaId);
    }

    // TODO: C# TO JAVA CONVERTER: Java annotations will not correspond to .NET attributes:
    //ORIGINAL LINE: [TestMethod][Owner("jthunter")] public async Task RoundTripAsync()
    // TODO: C# TO JAVA CONVERTER: There is no equivalent in Java to the 'async' keyword:
    //ORIGINAL LINE: [TestMethod][Owner("jthunter")] public async Task RoundTripAsync()
    public final Task RoundTripAsync() {
        Address tempVar = new Address();
        tempVar.setStreet("300B Chocolate Hole");
        tempVar.setCity("Great Cruz Bay");
        tempVar.setState("VI");
        PostalCode tempVar2 = new PostalCode();
        tempVar2.setZip(00830);
        tempVar2.setPlus4(0001);
        tempVar.setPostalCode(tempVar2);
        Address tempVar3 = new Address();
        tempVar3.setStreet("1 Microsoft Way");
        tempVar3.setCity("Redmond");
        tempVar3.setState("WA");
        PostalCode tempVar4 = new PostalCode();
        tempVar4.setZip(98052);
        tempVar3.setPostalCode(tempVar4);
        Address[] addresses = { tempVar, tempVar3 };

        String sampleComment = "hello there";
        String sampleSDL = "some SDL";

        try (Stream stm = new MemoryStream()) {
            // Create a reusable, resizable buffer.
            //C# TO JAVA CONVERTER WARNING: Unsigned integer types have no direct equivalent in Java:
            //ORIGINAL LINE: MemorySpanResizer<byte> resizer = new MemorySpanResizer<byte>(RecordIOUnitTests
            // .InitialRowSize);
            MemorySpanResizer<Byte> resizer = new MemorySpanResizer<Byte>(RecordIOUnitTests.InitialRowSize);

            // Write a RecordIO stream.
            // TODO: C# TO JAVA CONVERTER: There is no equivalent to 'await' in Java:
            // TODO: C# TO JAVA CONVERTER: The following lambda contained an unresolved 'out' keyword - these are
            // not converted by C# to Java Converter:
            //C# TO JAVA CONVERTER WARNING: Unsigned integer types have no direct equivalent in Java:
            //ORIGINAL LINE: Result r = await stm.WriteRecordIOAsync(new Segment(sampleComment, sampleSDL), (long
            // index, out ReadOnlyMemory<byte> body) =>
            Result r = await
            RecordIOStream.WriteRecordIOAsync(stm,
                new Segment(sampleComment, sampleSDL), (long index, out ReadOnlyMemory<Byte>body) ->
            {
                body = null;
                if (index >= addresses.length) {
                    return Result.Success;
                }

                Out<ReadOnlyMemory<Byte>> tempOut_body = new Out<ReadOnlyMemory<Byte>>();
                Task tempVar5 = this.WriteAddress(resizer, addresses[index], tempOut_body);
                body = tempOut_body.get();
                return tempVar5;
            });

            // Read a RecordIO stream.
            ArrayList<Address> addressesRead = new ArrayList<Address>();
            stm.Position = 0;
            //C# TO JAVA CONVERTER WARNING: Unsigned integer types have no direct equivalent in Java:
            //ORIGINAL LINE: resizer = new MemorySpanResizer<byte>(1);
            resizer = new MemorySpanResizer<Byte>(1);
            // TODO: C# TO JAVA CONVERTER: There is no equivalent to 'await' in Java:
            r = await
            RecordIOStream.ReadRecordIOAsync(stm, record ->
            {
                assert !record.IsEmpty;

                Address obj;
                Out<Address> tempOut_obj = new Out<Address>();
                r = this.ReadAddress(record, tempOut_obj);
                obj = tempOut_obj.get();
                ResultAssert.IsSuccess(r);
                addressesRead.add(obj);
                return Result.Success;
            }, segment ->
            {
                assert !segment.IsEmpty;

                Segment obj;
                Out<Segment> tempOut_obj =
                    new Out<Segment>();
                r = this.ReadSegment(segment, tempOut_obj);
                obj = tempOut_obj.get();
                ResultAssert.IsSuccess(r);
                assert obj.Comment == sampleComment;
                assert obj.SDL == sampleSDL;
                return Result.Success;
            }, resizer);

            ResultAssert.IsSuccess(r);

            // Check that the values all round-tripped.
            assert addresses.length == addressesRead.size();
            for (int i = 0; i < addresses.length; i++) {
                assert addresses[i] == addressesRead.get(i);
            }
        }
    }

    //C# TO JAVA CONVERTER WARNING: Unsigned integer types have no direct equivalent in Java:
    //ORIGINAL LINE: private Result ReadAddress(Memory<byte> buffer, out Address obj)
    private Result ReadAddress(Memory<Byte> buffer, Out<Address> obj) {
        RowBuffer row = new RowBuffer(buffer.Span, HybridRowVersion.V1, this.resolver);
        Reference<RowBuffer> tempReference_row =
            new Reference<RowBuffer>(row);
        RowReader reader = new RowReader(tempReference_row);
        row = tempReference_row.get();

        // Use the reader to dump to the screen.
        Reference<RowReader> tempReference_reader =
            new Reference<RowReader>(reader);
        String str;
        // TODO: C# TO JAVA CONVERTER: The following method call contained an unresolved 'out' keyword - these
        // cannot be converted using the 'Out' helper class unless the method is within the code being modified:
        Result r = DiagnosticConverter.ReaderToString(tempReference_reader, out str);
        reader = tempReference_reader.get();
        if (r != Result.Success) {
            obj.setAndGet(null);
            return r;
        }

        System.out.println(str);

        // Reset the reader and materialize the object.
        Reference<RowBuffer> tempReference_row2 =
            new Reference<RowBuffer>(row);
        reader = new RowReader(tempReference_row2);
        row = tempReference_row2.get();
        Reference<RowReader> tempReference_reader2 =
            new Reference<RowReader>(reader);
        Result tempVar = AddressSerializer.Read(tempReference_reader2, obj);
        reader = tempReference_reader2.get();
        return tempVar;
    }

    //C# TO JAVA CONVERTER WARNING: Unsigned integer types have no direct equivalent in Java:
    //ORIGINAL LINE: private Result ReadSegment(Memory<byte> buffer, out Segment obj)
    private Result ReadSegment(Memory<Byte> buffer, Out<Segment> obj) {
        RowBuffer row = new RowBuffer(buffer.Span, HybridRowVersion.V1, SystemSchema.LayoutResolver);
        Reference<RowBuffer> tempReference_row =
            new Reference<RowBuffer>(row);
        RowReader reader = new RowReader(tempReference_row);
        row = tempReference_row.get();

        // Use the reader to dump to the screen.
        Reference<RowReader> tempReference_reader =
            new Reference<RowReader>(reader);
        String str;
        // TODO: C# TO JAVA CONVERTER: The following method call contained an unresolved 'out' keyword - these cannot be converted using the 'Out' helper class unless the method is within the code being modified:
        Result r = DiagnosticConverter.ReaderToString(tempReference_reader, out str);
        reader = tempReference_reader.get();
        if (r != Result.Success) {
            obj.setAndGet(null);
            return r;
        }

        System.out.println(str);

        // Reset the reader and materialize the object.
        Reference<RowBuffer> tempReference_row2 = new Reference<RowBuffer>(row);
        reader = new RowReader(tempReference_row2);
        row = tempReference_row2.get();
        Reference<RowReader> tempReference_reader2 = new Reference<RowReader>(reader);
        Result tempVar = SegmentSerializer.Read(tempReference_reader2, obj.clone());
        reader = tempReference_reader2.get();
        return tempVar;
    }

    //C# TO JAVA CONVERTER WARNING: Unsigned integer types have no direct equivalent in Java:
    //ORIGINAL LINE: private Result WriteAddress(MemorySpanResizer<byte> resizer, Address obj, out
    // ReadOnlyMemory<byte> buffer)
    private Result WriteAddress(MemorySpanResizer<Byte> resizer, Address obj,
                                Out<ReadOnlyMemory<Byte>> buffer) {
        RowBuffer row = new RowBuffer(RecordIOUnitTests.InitialRowSize, resizer);
        row.initLayout(HybridRowVersion.V1, this.addressLayout, this.resolver);
        Reference<RowBuffer> tempReference_row =
            new Reference<RowBuffer>(row);
        Result r = RowWriter.WriteBuffer(tempReference_row, obj, AddressSerializer.Write);
        row = tempReference_row.get();
        if (r != Result.Success) {
            buffer.setAndGet(null);
            return r;
        }

        buffer.setAndGet(resizer.getMemory().Slice(0, row.length()));
        return Result.Success;
    }
}

File: src/Serialization/HybridRow.Tests.Unit/RecordIOUnitTests.cs
// ------------------------------------------------------------
//  Copyright (c) Microsoft Corporation.  All rights reserved.
// ------------------------------------------------------------

namespace Microsoft.Azure.Cosmos.Serialization.HybridRow.Tests.Unit
{
    using System;
    using System.Collections.Generic;
    using System.IO;
    using System.Runtime.InteropServices;
    using System.Threading.Tasks;
    using Microsoft.Azure.Cosmos.Serialization.HybridRow.IO;
    using Microsoft.Azure.Cosmos.Serialization.HybridRow.Layouts;
    using Microsoft.Azure.Cosmos.Serialization.HybridRow.RecordIO;
    using Microsoft.Azure.Cosmos.Serialization.HybridRow.Schemas;
    using Microsoft.Azure.Cosmos.Serialization.HybridRow.Tests.Unit.CustomerSchema;
    using Microsoft.Azure.Cosmos.Serialization.HybridRowGenerator;
    using Microsoft.VisualStudio.TestTools.UnitTesting;

    [TestClass]
    public class RecordIOUnitTests
    {
        private const int InitialRowSize = 0;

        [TestMethod]
        [Owner("jthunter")]
        public void LoadSchema()
        {
            LayoutResolver systemResolver = SchemasHrSchema.LayoutResolver;
            Layout segmentLayout = systemResolver.Resolve((SchemaId)SegmentHybridRowSerializer.SchemaId);
            Assert.AreEqual(segmentLayout.Name, "Segment");
            Assert.AreEqual(segmentLayout.SchemaId, (SchemaId)SegmentHybridRowSerializer.SchemaId);

            Layout recordLayout = systemResolver.Resolve((SchemaId)RecordHybridRowSerializer.SchemaId);
            Assert.AreEqual(recordLayout.Name, "Record");
            Assert.AreEqual(recordLayout.SchemaId, (SchemaId)RecordHybridRowSerializer.SchemaId);
        }

        [TestMethod]
        [Owner("jthunter")]
        public async Task RoundTripAsync()
        {
            Address[] addresses =
            {
                new Address
                {
                    Street = "300B Chocolate Hole",
                    City = "Great Cruz Bay",
                    State = "VI",
                    PostalCode = new PostalCode
                    {
                        Zip = 00830,
                        Plus4 = 0001,
                    },
                },
                new Address
                {
                    Street = "1 Microsoft Way",
                    City = "Redmond",
                    State = "WA",
                    PostalCode = new PostalCode
                    {
                        Zip = 98052,
                    },
                },
            };

            string sampleComment = "hello there";
            Namespace sampleNs = new Namespace { Name = "some sample namespace" };

            using (Stream stm = new MemoryStream())
            {
                // Create a reusable, resizable buffer.
                MemorySpanResizer<byte> resizer = new MemorySpanResizer<byte>();

                // Write a RecordIO stream.
                Result r = await stm.WriteRecordIOAsync(
                    new Segment(sampleComment, sampleNs),
                    (long index, out ReadOnlyMemory<byte> body) =>
                    {
                        body = default;
                        if (index >= addresses.Length)
                        {
                            return Result.Success;
                        }

                        return RecordIOUnitTests.WriteAddress(resizer, addresses[index], out body);
                    });

                ResultAssert.IsSuccess(r);

                // Read a RecordIO stream.
                List<Address> addressesRead = new List<Address>();
                stm.Position = 0;
                resizer = new MemorySpanResizer<byte>(1);
                r = await stm.ReadRecordIOAsync(
                    (ReadOnlyMemory<byte> record) =>
                    {
                        Assert.IsFalse(record.IsEmpty);

                        r = RecordIOUnitTests.ReadAddress(record, out Address obj);
                        ResultAssert.IsSuccess(r);
                        addressesRead.Add(obj);
                        return Result.Success;
                    },
                    segment =>
                    {
                        Assert.IsFalse(segment.IsEmpty);

                        r = RecordIOUnitTests.ReadSegment(segment, out Segment obj);
                        ResultAssert.IsSuccess(r);
                        Assert.AreEqual(obj.Comment, sampleComment);
                        Assert.AreEqual(obj.Schema.Name, sampleNs.Name);
                        return Result.Success;
                    },
                    resizer);

                ResultAssert.IsSuccess(r);

                // Check that the values all round-tripped.
                Assert.AreEqual(addresses.Length, addressesRead.Count);
                for (int i = 0; i < addresses.Length; i++)
                {
                    Assert.IsTrue(default(AddressHybridRowSerializer).Comparer.Equals(addresses[i], addressesRead[i]));
                }
            }
        }

        private static Result ReadSegment(ReadOnlyMemory<byte> buffer, out Segment obj)
        {
            RowReader reader = new RowReader(buffer, HybridRowVersion.V1, SchemasHrSchema.LayoutResolver);

            // Use the reader to dump to the screen.
            Result r = DiagnosticConverter.ReaderToString(ref reader, out string str);
            if (r != Result.Success)
            {
                obj = default;
                return r;
            }

            Console.WriteLine(str);

            // Reset the reader and materialize the object.
            // TODO: remove this cost-cast when ReadOnlyRowBuffer exists.
            // The cost-cast implied by MemoryMarshal.AsMemory is only safe here because:
            // 1. Only READ operations are performed on the row.
            // 2. The row is not allowed to escape this code.
            RowBuffer row = new RowBuffer(
                MemoryMarshal.AsMemory(buffer).Span,
                HybridRowVersion.V1,
                SchemasHrSchema.LayoutResolver);
            RowCursor root = RowCursor.Create(ref row);
            return default(SegmentHybridRowSerializer).Read(ref row, ref root, true, out obj);
        }

        private static Result WriteAddress(MemorySpanResizer<byte> resizer, Address obj, out ReadOnlyMemory<byte> buffer)
        {
            RowBuffer row = new RowBuffer(RecordIOUnitTests.InitialRowSize, resizer);
            Layout addressLayout = CustomerSchemaHrSchema.LayoutResolver.Resolve((SchemaId)AddressHybridRowSerializer.SchemaId);
            row.InitLayout(HybridRowVersion.V1, addressLayout, CustomerSchemaHrSchema.LayoutResolver);
            Result r = default(AddressHybridRowSerializer).Write(ref row, ref RowCursor.Create(ref row, out RowCursor _), true, default, obj);
            if (r != Result.Success)
            {
                buffer = default;
                return r;
            }

            buffer = resizer.Memory.Slice(0, row.Length);
            return Result.Success;
        }

        private static Result ReadAddress(ReadOnlyMemory<byte> buffer, out Address obj)
        {
            RowReader reader = new RowReader(buffer, HybridRowVersion.V1, CustomerSchemaHrSchema.LayoutResolver);

            // Use the reader to dump to the screen.
            Result r = DiagnosticConverter.ReaderToString(ref reader, out string str);
            if (r != Result.Success)
            {
                obj = default;
                return r;
            }

            Console.WriteLine(str);

            // Reset the reader and materialize the object.
            // TODO: remove this cost-cast when ReadOnlyRowBuffer exists.
            // The cost-cast implied by MemoryMarshal.AsMemory is only safe here because:
            // 1. Only READ operations are performed on the row.
            // 2. The row is not allowed to escape this code.
            RowBuffer row = new RowBuffer(
                MemoryMarshal.AsMemory(buffer).Span,
                HybridRowVersion.V1,
                CustomerSchemaHrSchema.LayoutResolver);
            return default(AddressHybridRowSerializer).Read(ref row, ref RowCursor.Create(ref row, out RowCursor _), true, out obj);
        }
    }
}
